File: laboratory/src/test/java/net/kolotyluk/loom/LagTests.java
package net.kolotyluk.loom;

import org.junit.jupiter.api.Test;

import kr.pe.kwonnam.slf4jlambda.LambdaLogger;
import kr.pe.kwonnam.slf4jlambda.LambdaLoggerFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.util.concurrent.Callable;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.Consumer;
import java.util.function.Supplier;

import static org.junit.jupiter.api.Assertions.*;

public class LagTests {
    // static final Logger logger1 = LoggerFactory.getLogger(LagTestsK.class);
    static final LambdaLogger logger = LambdaLoggerFactory.getLogger(LagTests.class);

    public record WithInterruptHandler(AtomicInteger value, Lag lag) implements Runnable {
        @Override
        public void run() {
            value.set(1);
            lag.sleep(cause -> value.set(2));
            // if (! Thread.interrupted()) value.set(2);
            if (value.get() == 1) value.set(3);
        }
    }

    public record WithoutInterruptHandler(AtomicInteger value, Lag lag) implements Runnable {
        @Override
        public void run() {
            var count = 0;
            try {
                value.set(1);
                for (int i = 0; i < 10; i++) {
                    count++;
                    logger.debug("i = {}", i);
                    logger.debug("Thread.currentThread().isInterrupted() = {}", () -> Thread.currentThread().isInterrupted());
                    if (Thread.currentThread().isInterrupted()) {
                        logger.debug("Interrupted value = {}", () -> value.get());
                        value.set(2);
                        logger.debug("            value = {}", () -> value.get());
                        break;
                    }
                    else lag.sleep();
                }
                if (value.get() == 1) value.set(3);
                logger.debug("count = {}", count);
            } catch (Exception e) {
                logger.debug("caught InterruptedException");
            }

        }
    }

}

File: laboratory/src/test/kotlin/net/kolotyluk/loom/LagTests.kt
package net.kolotyluk.loom

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertTrue

import org.junit.jupiter.api.Test
import java.lang.Thread.sleep

import java.time.Duration
import kotlin.concurrent.thread

class LagTests {

    private val minimumDuration = Duration.ofMillis(10)
    private val maximumDuration = Duration.ofMillis(20)

    @Test
    fun durationDefinite() {
        val lag = Lag(minimumDuration);
        val computedDuration = lag.duration
        assertEquals(minimumDuration, computedDuration)
    }

    @Test
    fun durationIndefinite() {
        val lag = Lag(minimumDuration, maximumDuration);
        val computedDuration = lag.duration
        assertNotEquals(minimumDuration, computedDuration)
        assertNotEquals(maximumDuration, computedDuration)
        assertTrue(computedDuration.minus(minimumDuration).isPositive)
        assertTrue(computedDuration.minus(maximumDuration).isNegative)
    }
}